Mathlib/Algebra/Ring/SumsOfSquares.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2024 Florent Schaffhauser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Florent Schaffhauser
+-/
+
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Submonoid.Basic
+import Mathlib.Algebra.Order.Ring.Defs
+
+/-!
+# Sums of squares
+
+We introduce sums of squares in a type `R` endowed with an `[Add R]`, `[Zero R]` and `[Mul R]`
+instances. Sums of squares in `R` are defined by an inductive predicate `isSumSq : R → Prop`:
+`0 : R` is a sum of squares and if `S` is a sum of squares, then for all `a : R`, `a * a + S` is a
+sum of squares in `R`.
+
+## Main declaration
+
+- The predicate `isSumSq : R → Prop`, defining the property of being a sum of squares in `R`.
+
+## Auxiliary declarations
+
+- The type `SumSqIn R` : in an additive monoid with multiplication `R`, we introduce the type
+`SumSqIn R` as the submonoid of `R` whose carrier is the subset `{S : R | isSumSq S}`.
+
+## Theorems
+
+- `isSumSq.add`  : if `S1` and `S2` are sums of squares in `R`, then so is `S1 + S2`.
+- `SumSq.nonneg` : in a linearly ordered semiring `R` with an `[ExistsAddOfLE R]` instance, sums of
+squares are non-negative.
+
+-/
+
+/--
+In a type `R` with an addition, a zero element and a multiplication, the property of being a sum of
+squares is defined by an inductive predicate: `0 : R` is a sum of squares and if `S` is a sum of
+squares, then for all `a : R`, `a * a + S` is a sum of squares in `R`.
+-/
+
+@[mk_iff]
+inductive isSumSq {R : Type*} [Add R] [Zero R] [Mul R] : R → Prop
+  | zero                              : isSumSq 0
+  | sq_add (a S : R) (pS : isSumSq S) : isSumSq (a * a + S)
+
+/--
+If `S1` and `S2` are sums of squares in a semiring `R`, then `S1 + S2` is a sum of squares in `R`.
+-/
+
+theorem isSumSq.add {R : Type*} [AddMonoid R] [Mul R] {S1 S2 : R} (p1 : isSumSq S1)
+    (p2 : isSumSq S2) : isSumSq (S1 + S2) := by
+  induction p1 with
+  | zero             => rewrite [zero_add]; exact p2
+  | sq_add a S pS ih => rewrite [add_assoc]; exact isSumSq.sq_add a (S + S2) ih
+
+/--
+In an additive monoid with multiplication `R`, the type `SumSqIn R` is the submonoid of sums of
+squares in `R`.
+-/
+
+def SumSqIn (R : Type*) [AddMonoid R] [Mul R] : AddSubmonoid R where
```

Comment:
Can you please add the fact that `SumSqIn` is the submonoid generated by the squares? It should be very easy and I think this is all we want to know about it.

Reply:
Sure, I'll do that asap and resubmit!